### PR TITLE
polish sparse interface, new hadamard product

### PR DESCRIPTION
**Pull request by GiggleLiu**

### CODE REVIEW

File: src/linalg.jl [hunk 1]
```diff
@@ -217,3 +218,32 @@ for FUNC in [:randn!, :rand!]
...
+
+
+########### Hadamard Product
+sparse_ranking(a::AbstractMatrix) = 6
+sparse_ranking(a::SDMatrix) = 5
+sparse_ranking(a::SDSparseMatrixCSC) = 4
+sparse_ranking(a::SDPermMatrix) = 3
+sparse_ranking(a::SDDiagonal) = 2
+sparse_ranking(a::IMatrix) = 1
+
```
Comment (Roger-luo) [inline src/linalg.jl]: docs?

File: src/linalg.jl [hunk 2]
```diff
@@ -217,3 +218,32 @@ for FUNC in [:randn!, :rand!]
         return m
     end
 end
+
+
+########### Hadamard Product
+sparse_ranking(a::AbstractMatrix) = 6
```
Comment (Roger-luo) [inline src/linalg.jl]: docs

File: src/staticize.jl [hunk 3]
```diff
@@ -52,3 +52,13 @@ dynamicize(A::PermMatrix) = PermMatrix(Vector(A.perm), Vector(A.vals))
 function dynamicize(A::SSparseMatrixCSC)
     SparseMatrixCSC(A.m, A.n, Vector(A.colptr), Vector(A.rowval), Vector(A.nzval))
 end
+
+function findnz(M::SDMatrix)
```
Comment (Roger-luo) [inline src/staticize.jl]: why not overload the one in `SparseArrays`?

File: src/linalg.jl (hunk 1 above)
Comment (GiggleLiu) [reply src/linalg.jl]: it is not exported.

Comment (GiggleLiu) [reply src/linalg.jl]: Merge it first, we can add docs later, new feature in YaoBlocks requires hadamard test.

File: src/staticize.jl (hunk 3 above)
Comment (GiggleLiu) [reply src/staticize.jl]: it can not be defined for Diagonal matrix due to pirating.

**Comment by GiggleLiu** [comment]
I will merge it for `YaoBlocks`, we can polish it latter.

**Pull request merged**